onnxruntime/core/platform/env.cc

```diff
@@ -34,4 +34,29 @@ std::ostream& operator<<(std::ostream& os, gsl::span<const LogicalProcessors> af
 
 Env::Env() = default;
 
+std::pair<int, std::string> GetErrnoInfo() {
+  auto err = errno;
+  std::string msg;
+
+  if (err != 0) {
+    char buf[512];
+
+#if defined(_WIN32)
+    auto ret = strerror_s(buf, sizeof(buf), err);
+    msg = ret == 0 ? buf : "Failed to get error message";  // buf is guaranteed to be null terminated by strerror_s
+#else
+    // strerror_r return type differs by platform.
+    auto ret = strerror_r(err, buf, sizeof(buf));
+    if (std::is_same_v<decltype(ret), int>) {  // POSIX returns int
```

Comment:
nit: if constexpr
```suggestion
    if constexpr (std::is_same_v<decltype(ret), int>) {  // POSIX returns int
```

Reply:
On Linux it highly depends on whether the _GNU_SOURCE macro is defined or not. For example, whether we are using "--std=c++17" or "--std=gnu++17". It is really very tricky. Thanks for providing such a so great solution.

Reply:
This is the approach used in absl so I take no credit for it.